### ISSUE TITLE
Possible regression in 0.0.24: `torch.utils.checkpoint` gradients through `scalar()` differ from 0.0.23

### ISSUE THREAD

**Issue by ZhuLin-612019** [ar4/deepwave#120]
# Possible regression in 0.0.24: `torch.utils.checkpoint` gradients through `scalar()` differ from 0.0.23

Hi, and thank you for Deepwave — it's been wonderful to work with, and the
checkpointing example in particular has been really helpful for keeping memory
under control on larger problems.

I think I may have run into a small regression and wanted to share it in case
it's useful. While using gradient checkpointing for FWI I noticed the inversion
had stopped making progress, and after switching Deepwave versions it looks like
something changed between **0.0.23 and 0.0.24**: when the propagation is run in
time segments with the wavefield state threaded through
`torch.utils.checkpoint` (the pattern in `docs/example_checkpointing.py`), the
gradient w.r.t. the velocity model comes out wrong on 0.0.24 and later, while
it's correct on 0.0.23. Everything else (PyTorch, CUDA, the script) was held
fixed across the versions.

I also noticed `docs/example_checkpointing.py` is affected the same way: on
0.0.23 it recovers the Marmousi structure as documented, but on
0.0.24 / 0.0.25 / 0.0.26 the "Out" panel comes out ≈ the initial model. I could
of course be misunderstanding something, so please feel free to correct me.

## Environment

- Deepwave: **0.0.22 / 0.0.23 look OK; 0.0.24 / 0.0.25 / 0.0.26 look broken** 
- PyTorch: 2.12.1+cu130
- Python: 3.10, Linux, CUDA

## What I'm seeing

In the minimal reproduction below, the checkpointed run gives a gradient whose
**norm collapses to ≈ 0** (`||g2|| / ||g1|| = 0.0000`), so the optimiser can't
make progress, whereas on 0.0.23 it matches the non-checkpointed gradient
exactly. The forward output is identical in both cases — only the backward
seems affected.

## Minimal reproduction (no external data)

```python
"""Minimal self-contained repro: torch.utils.checkpoint + scalar() gradient.

No external data. Runs the propagation in time segments, threading the wavefield
state (wavefield_0/m1, psi*, zeta*) from one segment to the next, and wraps the
non-final segments in torch.utils.checkpoint -- exactly the pattern in
docs/example_checkpointing.py. We compare grad(loss) w.r.t. the velocity model
for n_segments=1 (no checkpoint) vs n_segments=2 (first segment checkpointed).
Checkpointing must not change the gradient -- it only trades compute for memory.
"""

import torch
import torch.utils.checkpoint

import deepwave
from deepwave import scalar

device = "cuda" if torch.cuda.is_available() else "cpu"

ny, nx, dx = 60, 60, 5.0
nt, freq, dt0 = 300, 25.0, 0.006
max_vel = 1600.0
n_shots = 2
pml = 20  # deepwave default pml_width; wavefield_size is built from it

# Simple two-layer velocity model
v_true = 1500.0 * torch.ones(ny, nx, device=device)
v_true[30:] = 1600.0

source_locations = torch.zeros(n_shots, 1, 2, dtype=torch.long, device=device)
source_locations[..., 1] = 5
source_locations[:, 0, 0] = torch.tensor([15, 45])

receiver_locations = torch.zeros(n_shots, nx, 2, dtype=torch.long, device=device)
receiver_locations[..., 1] = 5
receiver_locations[:, :, 0] = torch.arange(nx).repeat(n_shots, 1)

# CFL upsampling (matches example_checkpointing.py); pass the interval too.
dt, step_ratio = deepwave.common.cfl_condition(dx, dx, dt0, max_vel)
print("step_ratio =", step_ratio)
source_amplitudes = deepwave.common.upsample(
    deepwave.wavelets.ricker(freq, nt, dt0, 1.5 / freq).repeat(n_shots, 1, 1),
    step_ratio,
).to(device)
nt_up = source_amplitudes.shape[-1]


def wrap(v, chunk, wf0, wfm1, psiy, psix, zetay, zetax):
    """One time segment; passes in and returns the 6 wavefield-state tensors."""
    return scalar(
        v, dx, dt,
        source_amplitudes=chunk,
        source_locations=source_locations,
        receiver_locations=receiver_locations,
        pml_freq=freq, max_vel=max_vel,
        wavefield_0=wf0, wavefield_m1=wfm1,
        psiy_m1=psiy, psix_m1=psix, zetay_m1=zetay, zetax_m1=zetax,
        model_gradient_sampling_interval=step_ratio,
    )


def propagate(v, n_segments):
    def z():
        return torch.zeros(n_shots, ny + 2 * pml, nx + 2 * pml, device=device)

    wf0, wfm1, psiy, psix, zetay, zetax = z(), z(), z(), z(), z(), z()
    rec = torch.zeros(n_shots, nx, nt_up, device=device)
    k = 0
    chunks = torch.chunk(source_amplitudes, n_segments, dim=-1)
    for i, c in enumerate(chunks):
        a = (v, c, wf0, wfm1, psiy, psix, zetay, zetax)
        if i == len(chunks) - 1:
            wf0, wfm1, psiy, psix, zetay, zetax, r = wrap(*a)
        else:
            wf0, wfm1, psiy, psix, zetay, zetax, r = (
                torch.utils.checkpoint.checkpoint(wrap, *a, use_reentrant=True)
            )
        rec[..., k:k + c.shape[-1]] = r
        k += c.shape[-1]
    return rec


def grad_of(n_segments):
    v = v_true.clone().requires_grad_(True)
    rec = propagate(v, n_segments)
    ((rec ** 2).sum()).backward()
    return v.grad.clone()


g1 = grad_of(1)  # single segment -> final -> no checkpoint (reference)
try:
    g2 = grad_of(2)  # first of two segments is checkpointed
except Exception as e:  # 0.0.24+ may raise here
    print("checkpoint path RAISED ->", type(e).__name__)
    print(str(e).splitlines()[0])
    raise SystemExit

cos = torch.nn.functional.cosine_similarity(
    g1.flatten(), g2.flatten(), dim=0
).item()
ratio = (g2.norm() / g1.norm()).item()
print(f"cosine(n_seg=1, n_seg=2) = {cos:+.4f}   (expected ~ +1.0)")
print(f"||g2|| / ||g1||          = {ratio:.4f}   (expected ~ 1.0)")
```

Output on **0.0.23**:

```
step_ratio = 5
cosine(n_seg=1, n_seg=2) = +1.0000   (expected ~ +1.0)
||g2|| / ||g1||          = 1.0000   (expected ~ 1.0)
```

Output on **0.0.26**:

```
step_ratio = 5
cosine(n_seg=1, n_seg=2) = -0.2072   (expected ~ +1.0)
||g2|| / ||g1||          = 0.0000   (expected ~ 1.0)
```

## What I'd expect

Since checkpointing only trades compute for memory, I'd expect it not to change
the gradient — i.e. cosine ≈ +1.0 and the ratio ≈ 1.0 on all versions, as on
0.0.23.

Thanks again for the library, and apologies if I've missed something obvious.
Happy to test patches or provide any further detail.

**Comment by ar4** [comment]
Dear @ZhuLin-612019 

Thank you for such an excellent bug report. Even though the cause was somewhat complicated, your detailed analysis and example made it much easier to find than it would have otherwise been. Thank you for investing so much time into it.

I hope that the issue is now fixed. I have pushed it to the branch [fix/checkpoint-gradient-regression-120](https://github.com/ar4/deepwave/tree/fix/checkpoint-gradient-regression-120). If the build and testing goes well, I plan to push it out as a new release within the next day or so.

**Comment by ar4** [comment]
Thank you again for reporting this. I have now released v0.0.27, which I hope fixes it. Please let me know if you encounter any other issues or have questions or suggestions.
